packages/devtools_app/lib/src/shared/common_widgets.dart

```diff
@@ -254,6 +248,166 @@ class ClearButton extends IconLabelButton {
         );
 }
 
+class EditableList extends StatefulWidget {
+  const EditableList({
+    required this.entries,
+    required this.label,
+    this.isRefreshing,
+    this.onEntryAdded,
+    this.onEntryRemoved,
+    this.onRefresh,
+  });
+
+  final ValueListenable<List<String>> entries;
+  final ValueListenable<bool>? isRefreshing;
+  final Function(String)? onEntryAdded;
+  final Function(String)? onEntryRemoved;
+  final Function()? onRefresh;
+  final String label;
+
+  @override
+  State<StatefulWidget> createState() => _EditableListState();
+}
+
+class _EditableListState extends State<EditableList> {
+  @override
+  void initState() {
+    super.initState();
+    textFieldController = TextEditingController();
+    widget.isRefreshing?.addListener(_stateRefresher);
+  }
+
+  late final TextEditingController textFieldController;
+  final FocusNode textFieldFocusNode = FocusNode();
+  void _stateRefresher() {
+    // trigger a state update when the refreshing state updates
+    setState(() {});
+  }
+
+  @override
+  void dispose() {
+    textFieldController.dispose();
+    widget.isRefreshing?.removeListener(_stateRefresher);
+    super.dispose();
+  }
+
+  void _addNewPubRootDirecory() {
+    final value = textFieldController.value.text.trim();
+    textFieldController.clear();
+    if (widget.onEntryAdded != null && value.isNotEmpty) {
+      textFieldController.clear();
+      widget.onEntryAdded!(value);
+    }
+    textFieldFocusNode.requestFocus();
+  }
+
+  Widget _removeDirectoryButton(VoidCallback onPressed) {
+    return Container(
+      height: defaultIconSize + denseSpacing,
+      child: inputDecorationSuffixButton(Icons.delete, onPressed),
+    );
+  }
+
+  @override
+  Widget build(BuildContext context) {
```

Comment:
this build method is too long - let's break up into multiple widgets. It would also be reasonable to put this widget into its own file shared/editable_list.dart since it is relatively large.